src/jit/morph.cpp

```diff
@@ -11416,6 +11416,20 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                 }
             }
 
+            // If gtOp1 is a GT_FIELD, we need to pass down the mac if
+            // its parent is GT_ADDR, since the address of the field
+            // is part of an ongoing address computation. Otherwise
+            // op1 represents the value of the field and so any address
+            // calculations it does are in a new context.
+            if ((op1->gtOper == GT_FIELD) && (tree->gtOper != GT_ADDR))
```

Comment:
> the case where we add some offset to a base pointer (typically a GC pointer) and the amount that we add might push us past the set of unmapped pages

Right, but we're not saying that we insert null checks for every instance of pointer arithmetic we can find without a fold-onto-able dereference, are we?  I presume this code is here to implement the null check for `ldflda` (but at the same time do the optimization to omit its explicit null check if we find that it's feeding a dereference that would be on the guard page if the object were null).

>The whole thing feels a bit off to me

I agree.  I think it would feel cleaner to insert null checks where the language requires them (ldflda) in the importer, and fold them as an optimization, or perhaps have the importer simply flag them (`GTF_ADDR_NEEDS_NULLCHECK` or something) and then this code could clearly be generating or omitting those null checks, rather than leaving us guessing at what contract it's implementing.